maas/resource_maas_boot_source_selection.go

```diff
@@ -190,3 +240,87 @@ func getBootSourceSelection(client *client.Client, bootSource int, id int) (*ent
 
 	return bootSourceSelection, nil
 }
+
+func awaitImportComplete(client *client.Client, os string, release string, arches []string, timeout time.Duration) error {
+	if err := client.BootResources.Import(); err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	result := retry.RetryContext(ctx, timeout, func() *retry.RetryError {
+		// Retrieve all the boot resources with type synced. These resources are created when importing from a boot source selection.
+		allResources, err := getBootResources(client, "synced")
+		if err != nil {
+			return retry.NonRetryableError(err)
+		}
+
+		// Confirm that for all user provided architectures in the boot source selection there is at least one boot resource found.
+		// If any architecture is missing, then the import is still ongoing. To perform this operation, the usage of sets is required
+		// since users can provide duplicates and MAAS can accept them.
+		archesSet := set.From(arches)
+		archesFound := set.New[string](0)
+
+		for _, resource := range allResources {
+			if resource.Name == fmt.Sprintf("%s/%s", os, release) {
+				archesFound.Insert(strings.Split(resource.Architecture, "/")[0])
+
+				// To verify that all boot resources are fully synced, we need to confirm the `completed` status of all their
+				// individual files. This information is not accessible by the list read operation. Instead, we have to fetch
+				// the resources individually.
+				resourceDetails, err := client.BootResource.Get(resource.ID)
+				if err != nil {
+					return retry.NonRetryableError(err)
+				}
+
+				// For each boot resource the response contains a list of sets, with each set representing a version of the boot
+				// resource. e.g., 24.04-ga-24.04-20250424. Each set has a complete boolean flag which, if set, represents the
+				// completion of the image synchronization. If not, that means that the import is still ongoing.
+				for _, resourceSet := range resourceDetails.Sets {
+					if !resourceSet.Complete {
+						return retry.RetryableError(fmt.Errorf("image still importing, waiting... "))
+					}
+				}
+			}
+		}
+
+		// There is still difference between user selected architectures set and unique boot resource architectures for the given
+		// os/release. The import is still ongoing.
+		if !archesSet.Equal(archesFound) {
```

Comment:
very much appreciate all the comments above, this is a non-trivial bit of logic to follow